File: source/HttpInterface/WaveServerPageDirectory.cpp
/***************************************************************************
 *   Copyright (C) 2005-2011 Vidyasagara Guntaka                           *
 *   All rights reserved.                                                  *
 *   Author : Vidyasagara Reddy Guntaka                                    *
 ***************************************************************************/

#include "HttpInterface/WaveServerPageDirectory.h"
#include <Framework/Utils/AssertUtils.h>
#include "Framework/Utils/TraceUtils.h"
#include <Framework/Core/Wave.h>
#include "HttpInterface/WaveServerPage.h"
#include "HttpInterface/WaveServerPageDirectoryEntry.h"

namespace WaveNs
{

map<string, WaveServerPage *> WaveServerPageDirectory::m_serverPageMap;
WaveServerPageDirectoryEntry  WaveServerPageDirectory::m_waveServerPageDirectoryRoot ("/", NULL);

WaveServerPageDirectory::WaveServerPageDirectory ()
{
}

WaveServerPageDirectory::~WaveServerPageDirectory ()
{
}

bool WaveServerPageDirectory::isAKnownServerPage (const string& path)
{
    map<string, WaveServerPage *>::const_iterator element    = m_serverPageMap.find (path);
    map<string, WaveServerPage *>::const_iterator endElement = m_serverPageMap.end  ();

    if (endElement != element)
    {
        return (true);
    }
    else
    {
        return (false);
    }
}

void WaveServerPageDirectory::registerServerPage (WaveServerPage *pWaveServerPage)
{
    prismAssert (NULL != pWaveServerPage, __FILE__, __LINE__);

    string path             = pWaveServerPage->getPath ();
    bool   isAnExistingPage = isAKnownServerPage (path);

    m_waveServerPageDirectoryRoot.addSibling (path, pWaveServerPage);

    if (true == isAnExistingPage)
    {
        trace (TRACE_LEVEL_FATAL, "WaveServerPageDirectory::registerServerPage : A Page with the path \"" + path + "\" has already been registered.");

        prismAssert (false, __FILE__, __LINE__);
    }
    else
    {
        if (NULL != pWaveServerPage)
        {
            trace (TRACE_LEVEL_DEVEL, "WaveServerPageDirectory::registerServerPage : Registering a Wave Server Page : \"" + path + "\"");

            m_serverPageMap[path] = pWaveServerPage;
        }
    }
}

void WaveServerPageDirectory::registerServerPage (const string &path, WaveServerPage *pWaveServerPage)
{
    bool isAnExistingPage = isAKnownServerPage (path);

    m_waveServerPageDirectoryRoot.addSibling (path, pWaveServerPage);

    if (true == isAnExistingPage)
    {
        trace (TRACE_LEVEL_FATAL, "WaveServerPageDirectory::registerServerPage : A Page with the path \"" + path + "\" has already been registered.");

        prismAssert (false, __FILE__, __LINE__);
    }
    else
    {
        if (NULL != pWaveServerPage)
        {
            trace (TRACE_LEVEL_DEVEL, "WaveServerPageDirectory::registerServerPage : Registering a Wave Server Page : \"" + path + "\"");

            m_serverPageMap[path] = pWaveServerPage;
        }
    }
}

WaveServerPage *WaveServerPageDirectory::getWaveServerPage (const string &path)
{
    return (m_waveServerPageDirectoryRoot.getWaveServerPageForRelativePath (path));
}

ResourceId WaveServerPageDirectory::print (UI32 argc, vector<string> argv)
{
    m_waveServerPageDirectoryRoot.print ();

    return (SHELL_OK);
}

}

File: Java/source/com/CxWave/Wave/HttpInterface/WaveServerPageDirectory.java
/*************************************************************************************************************************
 * Copyright (C) 2015-2016 Vidyasagara Guntaka & CxWave, Inc * All rights reserved. * Author : Vidyasagara Reddy Guntaka *
 *************************************************************************************************************************/

package com.CxWave.Wave.HttpInterface;

import java.util.Map;
import java.util.Vector;

import com.CxWave.Wave.Framework.Utils.Assert.WaveAssertUtils;
import com.CxWave.Wave.Framework.Utils.Trace.WaveTraceUtils;
import com.CxWave.Wave.Shell.ShellDebug;
import com.CxWave.Wave.Shell.Annotations.ShellCommand;

public class WaveServerPageDirectory
{
    static Map<String, WaveServerPage>  m_serverPageMap;
    static WaveServerPageDirectoryEntry m_waveServerPageDirectoryRoot = new WaveServerPageDirectoryEntry ("/", null);

    public static boolean isAKnownServerPage (final String path)
    {
        return (m_serverPageMap.containsKey (path));
    }

    public static void registerServerPage (final WaveServerPage waveServerPage)
    {
        WaveAssertUtils.waveAssert (null != waveServerPage);

        final String path = waveServerPage.getPath ();
        final boolean isAnExistingPage = isAKnownServerPage (path);

        m_waveServerPageDirectoryRoot.addSibling (path, waveServerPage);

        if (true == isAnExistingPage)
        {
            WaveTraceUtils.fatalTracePrintf ("WaveServerPageDirectory.registerServerPage : A Page with the path \"%s\" has already been registered.", path);

            WaveAssertUtils.waveAssert ();
        }
        else
        {
            if (null != waveServerPage)
            {
                WaveTraceUtils.develTracePrintf ("WaveServerPageDirectory.registerServerPage : Registering a Wave Server Page : \"%s\"", path);

                m_serverPageMap.put (path, waveServerPage);
            }
        }
    }

    public static void registerServerPage (final String path, final WaveServerPage waveServerPage)
    {
        final boolean isAnExistingPage = isAKnownServerPage (path);

        m_waveServerPageDirectoryRoot.addSibling (path, waveServerPage);

        if (true == isAnExistingPage)
        {
            WaveTraceUtils.fatalTracePrintf ("WaveServerPageDirectory::registerServerPage : A Page with the path \"%s\" has already been registered.", path);

            WaveAssertUtils.waveAssert ();
        }
        else
        {
            if (null != waveServerPage)
            {
                WaveTraceUtils.develTracePrintf ("WaveServerPageDirectory.registerServerPage : Registering a Wave Server Page : \"%s\"", path);

                m_serverPageMap.put (path, waveServerPage);
            }
        }
    }

    public static WaveServerPage getWaveServerPage (final String path)
    {
        return (m_waveServerPageDirectoryRoot.getWaveServerPageForRelativePath (path));
    }

    @ShellCommand (shell = ShellDebug.class, briefHelp = "Prints Wave Server Page Tree")
    public static void printWaveServerPageTree (final Vector<String> arguments)
    {
        m_waveServerPageDirectoryRoot.print ();
    }
}
